server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -868,6 +868,7 @@ public IndexResult index(Index index) throws IOException {
                     indexResult = plan.earlyResultOnPreFlightError.get();
                     assert indexResult.getResultType() == Result.Type.FAILURE : indexResult.getResultType();
                 } else if (plan.indexIntoLucene || plan.addStaleOpToLucene) {
+                    localCheckpointTracker.advanceMaxSeqNo(plan.seqNoForIndexing);
```

Comment:
Properly setting the max sequence number before indexing might not only be a useful property for Lucene-indexed stuff, but in general. I would prefer not to hide it in this sub-condition.
It bugs me that the planning process is mutating the state of the LocalCheckpointTracker in this way when being primary, but not when being replica. If we want the flow for the primary and replica to be the same, we could do this instead as part of the planning process in `planIndexingAsNonPrimary`. This will make it so that both primary and replica manage this property in the planning phase (will also require adaptation to FollowingEngine).
As a follow-up, I think we should also investigate whether we can remove this state mutation from the planning process, i.e., not have the planning process assign a sequence number or change the max seq no, but do it as an explicit step here in the `index` method, given that it is such a fundamental step in processing a request.